utils/disk_space_monitor.cc

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright (C) 2024-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+#include <seastar/core/reactor.hh>
+#include <seastar/core/sleep.hh>
+#include <seastar/core/thread.hh>
+
+#include "utils/disk_space_monitor.hh"
+#include "utils/assert.hh"
+#include "utils/log.hh"
+
+using namespace std::chrono_literals;
+
+namespace utils {
+
+seastar::logger dsmlog("disk_space_monitor");
+
+disk_space_monitor::disk_space_monitor(abort_source& as, std::filesystem::path data_dir, config cfg)
+    : _as_sub(as.subscribe([this] () noexcept { _as.request_abort(); }))
+    , _data_dir(std::move(data_dir))
+    , _cfg(std::move(cfg))
+{}
+
+disk_space_monitor::~disk_space_monitor() {
+    SCYLLA_ASSERT(_poller_fut.available());
+}
+
+future<> disk_space_monitor::start() {
+    co_await get_filesystem_stats();
+    _poller_fut = poll();
+}
+
+future<> disk_space_monitor::stop() noexcept {
+    _as.request_abort();
+    co_await std::exchange(_poller_fut, make_ready_future());
+}
+
+disk_space_monitor::signal_connection_type disk_space_monitor::listen(signal_callback_type callback) {
+    return _signal_source.connect([this, callback = std::move(callback)] () mutable {
+        // the signal is fired inside thread context
+        callback(*this).get();
+    });
+}
+
+future<> disk_space_monitor::poll() {
+    return async([this] {
+        try {
+            while (!_as.abort_requested()) {
+                auto now = clock_type::now();
+                get_filesystem_stats().get();
+
+                _signal_source();
+
```

Comment:
Can you explain how signals2 requires a thread?

Reply:
`boost::signals2::signal_type` can't return a `future<>`.
But that said, we can call the subscribed callback under a gate (or phased barrier) and wait for it in stop().
I'll do that and get rid of the thread.

Reply:
Dropped async thread in v12 (5620073786)